File: Best Time to Buy and Sell Stock/BestTimeToBuyAndSellStock.java
/**
time complexity: O(n) where n is the size of the prices array
space complexity: O(1)
approach: to register a maximum profit, we need to sell a stock on the ith day. we must also have purhcased the stock on the day between 1st day and (i-1)th day when its price was minimum*/
class Solution {
    public int maxProfit(int[] prices) {
        int minPrice = prices[0]; // we cannot sell stock on the 1st day, initializing minPrice to prices[0]
        int maxProfit = 0;
        int len = prices.length;
        for(int i=1;i<len;i++)
        {
            minPrice = Math.min(minPrice,prices[i-1]);
            int currPrice = prices[i];
            maxProfit = Math.max(maxProfit,currPrice-minPrice); // maxProfit will be updated only when the current day's stock price is greater than the minPrice
        }
        return maxProfit;
    }
}

File: Best Time to Buy and Sell Stock/best_time_to_buy_and_sell_stock.py
'''
https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

time complexity: O(n) where n is the size of the prices list
space complexity: O(1)
approach: when we sell on ith day, we must have bought on the day with minimum price from 1st to (i-1)th day to register a profit
we need to return the max running profit.
'''
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        min_price = prices[0] #we cannot sell on the 1st day, so initialize the price on the 1st day as the minimum price
        length = len(prices)
        max_profit = 0
        for i in range(1,length):
            min_price = min(min_price, prices[i-1]) #keeping track of the minimum stock price before the current day(i)
            max_profit = max(max_profit, prices[i] - min_price) #max_profit will be updated only when the current day's stock price is greater than the min_price
        return max_profit